internal/research/opentelemetry/instrumentation-tracing.md

```diff
@@ -0,0 +1,95 @@
+
+## Tracing Capabilities of the OpenTelemetry SDK
+
+The OpenTelemetry SDK can export traces using the OTLP protocol using GRPC or HTTP. The trace context can be propagated between processes using the W3C TraceContext or B3 headers.
+
+Traces can be exported to external backends (e.g., Jaeger) using the OpenTelemetry collector (see [deployment](deployment-tracing.md)).
+
+## Code Instrumentation
+
+To instrument a Golang application, a `tracer` has to be created first. The following function initializes a tracer that exports spans to an OpenTelemetry collector using the GRPC protocol.
+
+```
+func initTracer() (trace.Tracer, error) {
+	ctx := context.Background()
+	allOpts := []otlpgrpc.Option{
+		otlpgrpc.WithEndpoint("localhost:55680"),
+		otlpgrpc.WithInsecure(),
+	}
+
+	driver := otlpgrpc.NewDriver(allOpts...)
+	exporter, err := otlp.NewExporter(ctx, driver)
+	if err != nil {
+		return nil, err
+	}
+	// For the demonstration, use sdktrace.AlwaysSample sampler to sample all traces.
+	// In a production application, use sdktrace.ProbabilitySampler with a desired probability.
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.NewWithAttributes(semconv.ServiceNameKey.String("my-service"))),
+	)
+	defer func() { _ = tp.Shutdown(ctx) }()
+
+	pusher := controller.New(
+		processor.New(
+			simple.NewWithExactDistribution(),
+			exporter,
+		),
+		controller.WithExporter(exporter),
+		controller.WithCollectPeriod(5*time.Second),
+	)
+
+	err = pusher.Start(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle this error in a sensible manner where possible
+	defer func() { _ = pusher.Stop(ctx) }()
+	otel.SetTracerProvider(tp)
+	global.SetMeterProvider(pusher.MeterProvider())
+	propagator := propagation.NewCompositeTextMapPropagator(propagation.Baggage{}, propagation.TraceContext{}, b3.B3{})
+	otel.SetTextMapPropagator(propagator)
+	return tp.Tracer("reviews-go"), nil
+}
+```
+
+Individual code snippets can be instrumented with spans as follows:
+
+```
+var span trace.Span
+ctx := r.Context()
+ctx, span = h.tracer.Start(ctx, "myRequestHandler")
+defer span.End()
+...
+```
+
+Spans can be enhanced with additional events. Events can contain custom attributes. E.g.:
+
+```
+...
+span.AddEvent("Acquiring lock", trace.WithAttributes(label.Int("pid", 1234))
+mutex.Lock()
+```
+
+## Automatic Instrumentation
+
+OpenTelemetry provides a list of [packages](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation) that can instrument specific Golang packages automatically. E.g., `net/http`.
+
+## Findings
+
+* The trace context propagation feature lacks documentation. We were not able to propagate the B3 context that has been injected by Istio to the OpenTelemetry SDK to the own span.
```

Comment:
what is "the own span" ? Our span?

Reply:
Then span that has been created for the current request by the SDK. Changed the wording a bit to make this clear.